### ISSUE TITLE
to_parquet  creates file_scheme: ’hive’ files. You cannot back into a merged dask dataframe.

### ISSUE THREAD

**Issue by littlegreenbean33** [dask/dask#2805]
Hi,

I am transforming a large file into multiple (300) parquet files, via dd.to_parquet.

 I expected that, same as I can do with HDF, I could specify a glob  and merge all parquet files back into a mssive dd that I can process in parallel.

However, dask.to_parquet creates parquet files in hive mode ( dirs) instead of "simple". Then, dask is unable to merge  multiple parquet files in this format, back into a dataframe.

Would it be possible to pass the file_scheme: ‘simple’|’hive’ trough to the fastparquet writer?

Regards.

JC

**Comment by martindurant** [comment]
While there is no "read this set of directories" function - where each directory contains a parquet data-set - you can achieve this in two ways:

- glob to generate lists of data files across all of your directories, and pass this to read_parquet
- make a list of dataframes, one for each directory you want to read, and pass this to concat.

The first method would be much less efficient, because it will read metadata from every file you want to load and not use the metadata files already existing in the directories. The second method will involve slightly more code, but not much. Something like the following should suffice:

    import dask.dataframe as dd
    data_dirs = [...]  # or generate by glob or any other means
    dfs = [dd.read_parquet(d, ...) for d in data_dirs]
    df = dd.concat(dfs)

If the directories form a logical progression on the index, such as successive time-periods, then the index and divisions of the resulting dataframe should allow for rapid selection among the data.

**Comment by littlegreenbean33** [comment]
@martindurant  Thanks for this

As for the first solution, read_parquet will object to read a list of parquet  "stores" if in hive format. It seems it only accepts  them if  "stores" are in "simple" format. 

As for the second approach I believe the creation of dfs will blow away the memory of the small nodes I am using ( 8GB each).

Incidentally, I am shorter on RAM  than on disk thus.....after I finish my processing, I re-read all my  partitions back  into pandas using fastparquet. Then I  write the dataframes  back, (from pandas  rather than dask) forcing "simple" storage format.........Ignorance is a bliss....

I wonder if we should consider this closed. I am mindful that people coming from pandas will expect the parquet interface to behave same as hdf interface (  actually behaves the same in dask as in pandas). 

Would this be a  feature?....or shall I close it.

Thanks for your help.

JC

**Comment by martindurant** [comment]
"read_parquet will object to read a list of parquet "stores""

Yes, you would need to pass a list of all the data files contained within the various directories.

" I believe the creation of dfs will blow away the memory"

Indeed no, this only sets up a lazy prescription of the dataframe, as with other dask operations. The metadata files will be read immediately, but processing will load chunk-by-chunk, and not tend to fill up your RAM if the chunks are relatively small. Dask was built to deal with bigger-than-memory datasets before dealing with distribution across a cluster.

**Comment by littlegreenbean33** [comment]
Thanks, for this.

My results are inconsistent with your guidance.

Again, I always run these tests from a clean ipython session, distributed mode and a fresh, clean conda install on ubuntu 16.04.

- Yes, you would need to pass a list of all the data files contained within the various directories.

    This does not work on my system. Here is the trace

```

ddf=dd.read_parquet("/home/julian/kaggle/wsdm/parquet/user_logs_21_dir/*/*.parquet")

ValueError                                Traceback (most recent call last)
<ipython-input-20-d525a5123eb5> in <module>()
----> 1 ddf=dd.read_parquet("/home/julian/kaggle/wsdm/parquet/user_logs_21_dir/*/*.parquet")

~/anaconda3/envs/wsm2/lib/python3.5/site-packages/dask/dataframe/io/parquet.py in read_parquet(path, columns, filters, categories, index, storage_options, engine)
    289         return _read_fastparquet(fs, paths, file_opener, columns=columns,
    290                                  filters=filters,
--> 291                                  categories=categories, index=index)
    292     else:
    293         return _read_pyarrow(fs, paths, file_opener, columns=columns,

~/anaconda3/envs/wsm2/lib/python3.5/site-packages/dask/dataframe/io/parquet.py in _read_fastparquet(fs, paths, myopen, columns, filters, categories, index, storage_options)
     49 
     50     if len(paths) > 1:
---> 51         pf = fastparquet.ParquetFile(paths, open_with=myopen, sep=myopen.fs.sep)
     52     else:
     53         try:

~/anaconda3/envs/wsm2/lib/python3.5/site-packages/fastparquet/api.py in __init__(self, fn, verify, open_with, sep)
     82         if isinstance(fn, (tuple, list)):
     83             basepath, fmd = metadata_from_many(fn, verify_schema=verify,
---> 84                                                open_with=open_with)
     85             self.fn = sep.join([basepath, '_metadata'])  # effective file
     86             self.fmd = fmd

~/anaconda3/envs/wsm2/lib/python3.5/site-packages/fastparquet/util.py in metadata_from_many(file_list, verify_schema, open_with)
    164     else:
    165         raise ValueError("Merge requires all PaquetFile instances or none")
--> 166     basepath, file_list = analyse_paths(file_list, sep)
    167 
    168     if verify_schema:

~/anaconda3/envs/wsm2/lib/python3.5/site-packages/fastparquet/util.py in analyse_paths(file_list, sep)
    221     if len({tuple([p.split('=')[0] for p in parts[l:-1]])
    222             for parts in path_parts_list}) > 1:
--> 223         raise ValueError('Partitioning directories do not agree')
    224     for path_parts in path_parts_list:
    225         for path_part in path_parts[l:-1]:

ValueError: Partitioning directories do not agree 
```

same error as

    `ddf=dd.read_parquet("/home/julian/kaggle/wsdm/parquet/user_logs_21_dir/*/*")`

- I believe the creation of dfs will blow away the memory"
    ```
    dfs = [dd.read_parquet(d, ...) for d in data_dirs]
    df = dd.concat(dfs)
   ```
I get mixed results here.

```
    In [21]: ddfs=[dd.read_parquet(os.path.join("/home/julian/kaggle/wsdm/parquet/user_logs_21_dir/",x)) for x in os.listdir("/home/juli...: an/kaggle/wsdm/parquet/user_logs_21_dir/")]
    In [22]: ddf=dd.concat(ddfs,interleave_partitions=True)

    In [23]: ddf.head()
    Out[23]: date  num_25  num_50  num_75  num_985  num_100  num_unq  total_secs
idx                                                                          
    0   2015-04-27       1       1       0        0        0        2   97.411003
    In [24]: f=ddf.date.count().persist()
```

ddf creation process will not fire any errors. However, as you can see a ddf.head() returns just one line.... fishy


ddf.date.count().persist() will run ( It looks like) locally in my ipython session. It will run in single process/thread mode ( as per cpu usage, worker and schedulers do nothing ) and will  blow my RAM. Eventually deleting the objects and calling gc.collect() will not release RAM. I have to kill the ipython session and start from scratch.

However, if I follow exactly the same steps, but concatenating from the  same set of files/parquet store with "file_scheme: ‘simple’ all works fine.  Head() returns number of rows as expected and the count operation runs across the cluster without noticeable increment of ram usage on the ipython session. 

Same tests, using the same set of files but on hdf, with separate files per partition ( 1m rows) runs perfectly well and across the cluster. Same results and approximate time.

**Comment by martindurant** [comment]
> "concatenating from the same set of files/parquet store with "file_scheme: ‘simple’ all works fine"

Is there any reason that you don't use this method?

> "ValueError: Partitioning directories do not agree "

This no longer happens in the latest version of fastparquet. You may wish to look at your list of file-paths to see if indeed there is some obvious inconsistency with how the partition names/values are given.

> "ddf.head() returns just one line.... fishy"

Getting just one row in `head()` is not necessarily unexpected: if your first partition happens to contain only one row, you will get that. The "first" partition should be the first row-group of the first data-set you provide in the list to concat (you could do a `head()` on that one dataset).

> "persist() will run ( It looks like) locally in my ipython session."

This depends on whether you have a dask.distributed.Client() available or not. If yes, it will use that preferentially. Note that the behaviour of persist is different in the local and distributed schedulers.

**Comment by littlegreenbean33** [comment]
@martindurant  Thanks for your time. I have been using  pandas + ipyparallel successfully for the last couple of years. I gave at the time a try to dask but I wasn't successful. I am taking a work break and giving back a stab to dask to see if I can comfortably and confidently move my work to dask/distributed. To  cut a long story short, I am learning. Thus I am trying every single function/option and analyze the results. I am using a couple of kaggle data sets as a test bed.

The test file I am using is a 400M rows .csv.gz  file. I have spitted into 1M row files for the sake of partitioning with unix split, which creates a set xaa...xot files.  In my opinion the 1-row head() remains fishy. e.g parquet-"simple" would give correct results

All my tests are with dask.distributed with a handful of equally configured ubuntu laptops using conda. I am hesitant to mix versions out of conda, as I may not be skillful enough to correctly understand and set dependencies. I moved away from a previous mixed windows/unix envs for the sake of reliability.

My understanding of persist, in this context, is that whatever calculations, intermediate results or final result remains in the distributed memory. Only when I trigger compute, results are back to my ipython session in the form of standard pandas dataframes/series. Even being very careful about the topic of persist (... still, learning and prone to make mistakes...) my ipython session quickly takes 3-4 gigabytes of memory which is making tricky to run the scheduler, a worker and my ipython session in the largest node of my minicluster. Need to keep researching about what I am doing incorrectly. I expected my ipython memory footprint to remain pretty much flat in absence of fat fingers...

Cheers and thanks for your feedback

JC

**Comment by martindurant** [comment]
You are right that persist, in the distributed scheduler context, will perform the computation you asked for asynchronously, and store the result in the memory of one of the workers; whereas compute will bring that result back to your controlling session (see also client.gather).

I do not, therefore, know what might be causing your memory usage to spike. The client process does have some work to do in analyzing the metadata and apportioning the work. 400 x 1M rows sounds like too fine-grained partitioning, so perhaps you can try with >>10M rows per row-group?

From what you had said earlier, I thought there was also partitioning by a variable going on, is this not the case? Then, the total number of row groups would be much higher (multiply by the number of possible values of variable(s) you chose), and this could be the difference between 'hive' and 'simple' modes, since you cannot partition at all with simple.

**Comment by littlegreenbean33** [comment]
@martindurant  Thanks for your comments.

I deleted my comments on memory consumption from the post. I figure out eventually that it wasn't my python session consuming the memory but the actual python process acting as worker ( I run my cli session, a scheduler and 1 worker --proc 1 in the same corei7/8GB while the rest of the workers run on a handful of corei5/8GB worker --proc 2). I got confused about what PID was  grabbing what.

 Why the worker on the main node  quickly takes 5-6GB of RAM and the never releases it, for the simple descriptive statistics/queries/groupby  I am doing, I do not know. Will try to figure out what am I doing wrong with a bit more testing.

I went to 1M rows mindful of the memory issues I was having. Will perform the test you mention along the week and will report back results.

The partition process I do is simple, big fat csv file into 1M rows files, reread, make a join to create a int32 column, make it index, save to parquet including index.
